src/bidiMapper/domains/script/Realm.ts

```diff
@@ -833,16 +812,176 @@ export class Realm {
+          },
         },
-      },
-      this.browsingContextId
+        browsingContext.id
+      );
+    }
+  }
+}
+
+export class WindowRealm extends Realm {
+  readonly #browsingContextId: BrowsingContext.BrowsingContext;
+  readonly #browsingContextStorage: BrowsingContextStorage;
+  readonly sandbox: string | undefined;
+
+  constructor(
+    browsingContextId: BrowsingContext.BrowsingContext,
+    browsingContextStorage: BrowsingContextStorage,
+    cdpClient: CdpClient,
+    eventManager: EventManager,
+    executionContextId: Protocol.Runtime.ExecutionContextId,
+    logger: LoggerFn | undefined,
+    origin: string,
+    realmId: Script.Realm,
+    realmStorage: RealmStorage,
+    sandbox: string | undefined,
+    sharedIdWithFrame: boolean
+  ) {
+    super(
+      browsingContextStorage,
+      cdpClient,
+      eventManager,
+      executionContextId,
+      logger,
+      origin,
+      realmId,
+      realmStorage,
+      sharedIdWithFrame
+    );
+
+    this.#browsingContextStorage = browsingContextStorage;
+    this.#browsingContextId = browsingContextId;
+    this.sandbox = sandbox;
+
+    this.initialize();
+  }
+
+  get browsingContext(): BrowsingContextImpl {
+    return this.#browsingContextStorage.getContext(this.#browsingContextId);
+  }
+
+  override get associatedBrowsingContexts(): [BrowsingContextImpl] {
+    return [this.browsingContext];
+  }
+
+  override get realmType(): 'window' {
+    return 'window';
+  }
+
+  override get realmInfo(): Script.WindowRealmInfo {
+    return {
+      ...this.baseInfo,
+      type: this.realmType,
+      context: this.#browsingContextId,
+      sandbox: this.sandbox,
+    };
+  }
+
+  override async evaluate(
+    expression: string,
+    awaitPromise: boolean,
+    resultOwnership: Script.ResultOwnership,
+    serializationOptions: Script.SerializationOptions,
+    userActivation?: boolean
+  ): Promise<Script.EvaluateResult> {
+    await this.#browsingContextStorage
+      .getContext(this.#browsingContextId)
+      .targetUnblockedOrThrow();
+
+    return await super.evaluate(
+      expression,
+      awaitPromise,
+      resultOwnership,
+      serializationOptions,
+      userActivation
     );
   }
+
+  override async callFunction(
+    functionDeclaration: string,
+    thisLocalValue: Script.LocalValue,
+    argumentsLocalValues: Script.LocalValue[],
+    awaitPromise: boolean,
+    resultOwnership: Script.ResultOwnership,
+    serializationOptions: Script.SerializationOptions,
+    userActivation?: boolean
+  ): Promise<Script.EvaluateResult> {
+    await this.#browsingContextStorage
+      .getContext(this.#browsingContextId)
+      .targetUnblockedOrThrow();
+
+    return await super.callFunction(
+      functionDeclaration,
+      thisLocalValue,
+      argumentsLocalValues,
+      awaitPromise,
+      resultOwnership,
+      serializationOptions,
+      userActivation
+    );
+  }
+}
+
+export class DedicatedWorkerRealm extends Realm {
```

Comment:
Move to a separate file please

Reply:
See other comment.

Reply:
Resolving to continue conversation in https://github.com/GoogleChromeLabs/chromium-bidi/pull/1709/files#r1453290301